rewrite whole code, waiting for element, timestamps in pgn

zeraye opened the pull request:
Whole code rewritten from scratch.

Changed:
 - `waitForElement` waits 5 seconds for element, if not program stops
 - automated clicking with `clickElement`
 - added timestamps to pgn
 - new endpoint in `show_matches`

Gelio commented on background.js:
```diff
@@ -1,49 +1,53 @@
+  let timeLeft = 1000;
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise(r => setTimeout(r, 100));
+    [isElement] = await browser.tabs.executeScript({code: `document.querySelector("${querySelector}") !== null;`});
+    timeLeft -= 100;
```
Minor:

You could extract the magic `100` to a single variable to ensure that the timeout delay and the value that is subtracted from `timeLeft` are always the same. In larger programs, this would also have an additional benefit that it shows that this "100" is the same as the "100" below, meaning you can click "Find all references" for a variable and see where this variable matters. Moreover, it shows that all of these places should change at the same time when the variable changes - when using a variable you only need to update one place and there is no risk that you update the timeout but not the `timeLeft`

```ts
const retryDelay = 100;

await new Promise(r => setTimeout(r, retryDelay));
// ...
timeLeft -= retryDelay;
```

Gelio commented on background.js:
```diff
@@ -1,49 +1,53 @@
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise(r => setTimeout(r, 100));
+    [isElement] = await browser.tabs.executeScript({code: `document.querySelector("${querySelector}") !== null;`});
```
Minor:

The code snippet that checks if the element is present is repetitive and prone to typos (the contents of the `code` parameter are not type-checked or validated in any way and will only cause errors during runtime).

You could ensure that the same `code` is used here and on line 9 by extracting it to a variable, e.g.:

```ts
const elementFoundScript = `document.querySelector(${querySelector}) !== null;`;
```

Gelio commented on background.js:
```diff
@@ -1,49 +1,53 @@
+  `});
+}
+
+const waitForElement = async querySelector => {
```
Very elegant solution overall :smile: I like it! Using `await` makes it a breeze to read :smile: Nice work

Gelio commented on background.js:
```diff
@@ -1,49 +1,53 @@
-      code: `!document.getElementById("form3-analyse").disabled;`
-    });
+    if (await waitForElement("[name='analyse']")) {
+      const [logged] = await browser.tabs.executeScript({code: `!document.querySelector("[name='analyse']").disabled;`});
```
The word _logged_ is more connected to _logging_ (storing log messages from a run of the program, for example to the console) than authentication. I assume this is meant to be connected to authentication. In that case, I suggest using the name `loggedIn` or `signedIn`, which is a more standard name related to auth :smile: 

```suggestion
      const [loggedIn] = await browser.tabs.executeScript({code: `!document.querySelector("[name='analyse']").disabled;`});
```

Gelio commented on background.js:
```diff
@@ -1,49 +1,53 @@
-  await browser.tabs.executeScript({
-    code: `document.getElementsByClassName("board-tab-item-underlined-component share-menu-tab-selector-tab")[0].click();`,
-  });
+browser.pageAction.onClicked.addListener(async () => {
```
Overall, I see there is very little code related to error handling :smile: If you want to make the extension more robust, you could think how to add error handling and show the user that something failed (e.g. there was a timeout in finding some element on the page, which could suggest the user has no Internet or the class names used on the page changed)

Gelio reviewed (approved):
Looks good to me overall :+1: I like the concise implementation of waiting for an element to appear :smile: Good job

Gelio reviewed (commented):
Your "small refactoring" changed the entire background.js script :smile: 

I haven't checked the code yet. I have some meta comments about commit messages: 

![image](https://user-images.githubusercontent.com/889383/158957457-574f7687-e342-4527-a15b-59f9beb610bb.png)


The _Small refactoring_ one does not immediately say what happened in the code. The commit header should describe the general changes that the commit introduces (the "what"). The reason (the fact it was done as a response to review comments) may (but does not have to) come in the body of the commit message. I said it does not have to be even included, because if you'll be looking at the commit history 1 year from now, the fact that it was done as a response to a PR review will probably be insignificant.

A rule to follow is: what information do you think will be useful when looking at the commit history 3 months/a year from now? Seeing the summary of the changes in the commit log helps (at least it helps me :smile: )

I suppose a more appropriate title for that commit would be

> Add basic error handling
>
> Show a message when the element cannot be found on the page.

The second commit message does not include a verb. "Debugging leftovers" is ambiguous and it is not immediately clear to me the commit removes them :smile: I would suggest rewording it to:

> Remove debugging leftovers

or, even better, fixup the commit with the previous one, as having a commit that removes some `console.log`s does not add much value in itself over not having those `console.log`s in the first place :smile: 

All in all, I usually recommend [this guide on writing commit messages](https://cbea.ms/git-commit/) as I usually see commits using that format. [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) are also a good convention to follow

zeraye commented:
I think these commits don't change that much. I changed one variable name and replaced some repeatable string with variable.
The biggest change were error messages, which I should separate in the second commit.

Gelio commented:
> I think these commits don't change that much. I changed one variable name and replaced some repeatable string with variable

You are right, my bad. I was viewing the diff for the whole PR instead of the diff for just that one commit. It didn't change that much, so that can still qualify as a "small refactor"

Gelio commented on background.js:
```diff
@@ -1,49 +1,82 @@
-  });
+const showMessage = async message => {
+  await browser.tabs.executeScript({code: `
+    closeBtn = document.createElement("span");
```
Minor:

Instead of creating a global variable, consider wrapping this code in [an IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) to use a regular variable declaration :smile:  The same goes for `alertBox`

Gelio commented on background.js:
```diff
@@ -1,49 +1,82 @@
+const showMessage = async message => {
+  await browser.tabs.executeScript({code: `
+    closeBtn = document.createElement("span");
+    closeBtn.textContent = "×";
```
Nice unicode :x: :smile:

Gelio commented on background.js:
```diff
@@ -1,49 +1,82 @@
+    closeBtn = document.createElement("span");
+    closeBtn.textContent = "×";
+    closeBtn.setAttribute("style", "font-weight:bold;float:right;font-size:20px;line-height:18px;cursor:pointer;padding-left:15px;");
+    closeBtn.setAttribute("onclick", "this.parentElement.style.display='none';");
```
This is certainly a non-standard way to attach a `click` handler :smile: The idiomatic way to do it would be:

```suggestion
    closeBtn.addEventListener('click', (event) => {
      alertBox.parentElement.removeChild(alertBox);
    });
```

Note that this also changes the way the element is removed - it is removed from the DOM rather than keeping the element hidden. Theoretically, if there would be hundreds of messages, they would stay in the DOM unnecessarily, increasing memory usage and slowing the page down. If we don't need the element anymore and we are not animating it disappear (e.g. with [a CSS `transition`](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Transitions/Using_CSS_transitions)), it should be safe to remove it

Gelio commented on background.js:
```diff
@@ -1,49 +1,82 @@
 
-  await new Promise(r => setTimeout(r, 500));
+  if (!isElement)
+    showMessage(`Cannot find \`${querySelector}\` element!`);
```
Minor:

If you return a rejected promise if there is an unrecoverable failure (e.g. an element cannot be found), this will stop the promise chain and next operations won't be executed. This is a way to stop executing the rest of the code if there was a failure. See <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#error_propagation> for more information

Gelio commented on background.js:
```diff
@@ -1,49 +1,82 @@
-      code:
-        "document.getElementById('form3-pgn').value = `" + pgn + "`; document.getElementsByClassName('submit')[0].click();",
-    });
+      if (await waitForElement("[name='pgn']")) {
```
Notice how you are building a pyramid of doom here by indenting each level of `if` :smile: This makes the code harder to read since you need to be weary of what code is inside which `if` and if there is code in some `else` (there is none here, but I'm making a general point about such pyramids)

Consider:

1. Inverting `if` conditions and returning early instead of indenting

    ```ts
    if (!(await waitForElement("[name='pgn']")) {
      return;
    }
    
    // the rest
    ```

2. Using rejected promises to stop the code from continuing if there is an unrecoverable failure. See my other comment on line 36 for more details

Gelio reviewed (approved):
LGTM :tada: I appreciate adding error handling :+1: 

I left some minor, general comments. They do not have to be addressed in this PR, as IMO this is good enough to be merged and creates value in itself

zeraye commented:
Adjustments from this review will be implemented in the very next update.
